Enable multiple beneficiaries

rainbreak opened the pull request:
Implementing #1.

Implementation plan:
- [x] replace `Auction` field `address beneficiary` with `address[] beneficiaries`
- [x] add `Auction` field `uint[] payouts`
- [x] override auction init to accept old behaviour (single beneficiary) or new arrays
- [x] assert `payouts.length == beneficiaries.length`
- [x] assert `sum(payouts) == collection_limit` (i.e. only implement in forward part of auction for now)
- [x] assert `payouts[0] >= start_bid`
- [x] modify `settleExcessBuy` to partition new bid excess among beneficiaries
- [x] ~~think about UX of specifying all payouts. Should there be a way to infer the remaining payout, to make the forward auction easier to initialise? (The collection limit is 'infinite' == 2 *\* 256 - 1, so this has to be reasoned with).~~ (addressed by #2)
- [x] create initialiser for two-way auction
- [x] remove collection limit arg in case of two-way with payouts array
- [x] more multi beneficiary two-way tests

rainbreak committed:
misc whitespace

rainbreak committed:
add beneficiaries array to Auction

rainbreak committed:
add limits array to Auction

rainbreak committed:
accept new arrays in internal auction constructor

rainbreak committed:
multiple beneficiaries in forward auction constructor

rainbreak committed:
benefit / limit arrays must be equal length

rainbreak committed:
assertIncreasing(array) function

rainbreak committed:
limits array must be increasing

rainbreak committed:
lower limit must be >= start bid

rainbreak committed:
lower limit must be zero

consider limits as threshold-over-which-paid rather than
threshold-under-which-paid

nmushegian commented:
> assert limits strictly increasing

could we specify payout amounts and have the limit be implicitly defined by cumulative sum? Would remove this assertion requirement and make consuming the api easier - is it more complex to implement?

rainbeam commented:
That's a better idea. Possible to implement - will just convert to limits internally.

rainbreak committed:
implement multiple beneficiaries

rainbreak committed:
specify payouts instead of limits

rainbreak committed:
first payout must be greater than start bid

Having the first payout less than the start bid is possible but messy to
implement.

This does not apply to reverse auctions.

rainbeam commented:
Idea for UX:

`payouts` can be exactly one element shorter than `beneficiaries`. If this is the case, the remaining payout is inferred as `collection_limit - sum(payouts)`.

nmushegian commented:
I was thinking collection limit would be computed as the sum of the payouts in the UI

rainbeam commented:
The ux problem is really motivated by the forward auction with no limit as the collection limit is implicitly set to 'infinite'.

Specifying the collection limit as well as payouts in a twoway auction is redundant so I'll remove that.

rainbreak committed:
tests: use single newAuction method

rainbreak committed:
tests: comment arg names for newAuction

rainbreak committed:
don't name args in reverse constructor

Solidity currently doesn't support overload of named calls.

rainbreak committed:
tests: single newAuction method in ForwardSplitTest

rainbreak committed:
set reversed inside main constructor

rainbreak committed:
multiple beneficiaries in two-way constructor

rainbreak committed:
add math tests

rainbreak committed:
add math sum function

rainbreak committed:
tests: fix comments

rainbreak committed:
tests: use implied base

rainbreak committed:
use sum directly

rainbreak committed:
don't require collection limit when payouts given

rainbreak committed:
multiple payout tests

rainbeam commented:
All outstanding tasks done and all tests passing. Will merge unless you've anything to add @nmushegian?

To recap, this PR implements multiple beneficiaries for the forward and two-way auctions. 

- An array of beneficiary addresses and an array of payout amounts are passed to the newAuction constructor. 
- The new auction will pay auction takings to the beneficiaries in turn, up to their associated payout amount (which is denoted in the `buying` token).
- In the two-way auction the sum of the payouts gives the collection limit - the point at which the auction will transition from a forward to a reverse auction.
- In the forward auction the sum of the payouts must equal `2 ** 256 - 1` ('infinity')
- The first payout must be greater than the start bid.
- The reverse auction is unaffected by this PR.
